mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala

```diff
@@ -0,0 +1,799 @@
+            buffer.weightSum(index) += w
+          }
+          if (buffer.nnz != null) {
+            buffer.nnz(index) += 1
+          }
+        }
+      }
+
+    }
+
+    /**
+     * Merges other into buffer.
+     */
+    private def mergeInitializedBuffers(buffer: Buffer, other: Buffer): Unit = {
+      // Each buffer needs to be properly initialized.
+      require(buffer.n > 0 && other.n > 0, (buffer.n, other.n))
+      require(buffer.n == other.n, (buffer.n, other.n))
+      // Mandatory scalar values
+      buffer.totalWeightSquareSum += other.totalWeightSquareSum
+      buffer.totalWeightSum += other.totalWeightSum
+      buffer.totalCount += other.totalCount
+      // Keep the original weight sums.
+      val weightSum1 = if (buffer.weightSum == null) null else { buffer.weightSum.clone() }
+      val weightSum2 = if (other.weightSum == null) null else { other.weightSum.clone() }
+
+      // This sum is going to be used as a denominator. In order to guarantee that the
+      // division is well-defined, we add an epsilon to the zero coefficients.
+      // This is not going to change the value of the resul since the numerator will also be zero.
+      val weightSum: BV[Double] = if (weightSum1 == null) null else {
+        require(weightSum2 != null, s"buffer=$buffer other=$other")
+        val x = b(weightSum1) :+ b(weightSum2)
+        la.max(x, Double.MinPositiveValue)
+      }
+
+
+      // Since the operations are dense, we can directly use BLAS calls here.
+      val deltaMean: BV[Double] = if (buffer.mean != null) {
+        require(other.mean != null)
+        b(other.mean) :- b(buffer.mean)
+      } else { null }
+
+      if (buffer.mean != null) {
+        require(other.mean != null)
+        require(weightSum != null)
+        b(buffer.mean) :+= deltaMean :* (b(weightSum2) :/ weightSum)
+      }
+
+      if (buffer.m2n != null) {
+        require(other.m2n != null)
+        val w = (b(weightSum1) :* b(weightSum2)) :/ weightSum
+        val z = (deltaMean :* deltaMean) :* w
+        b(buffer.m2n) :+= b(other.m2n) :+ z
+      }
+
+      if (buffer.m2 != null) {
+        require(other.m2 != null)
+        b(buffer.m2) :+= b(other.m2)
+      }
+
+      if (buffer.l1 != null) {
+        require(other.l1 != null)
+        b(buffer.l1) :+= b(other.l1)
+      }
+
+      if (buffer.max != null) {
+        require(other.max != null)
+        b(buffer.max) := la.max(b(buffer.max), b(other.max))
+      }
+
+      if (buffer.min != null) {
+        require(other.min != null)
+        b(buffer.min) := la.min(b(buffer.min), b(other.min))
+      }
+
+      if (buffer.nnz != null) {
+        require(other.nnz != null)
+        bl(buffer.nnz) :+= bl(other.nnz)
+      }
+
+      if (buffer.weightSum != null) {
+        require(other.weightSum != null)
+        b(buffer.weightSum) :+= b(other.weightSum)
+      }
+    }
+  }
+
+  private case class MetricsAggregate(
+      requested: Seq[Metrics],
+      startBuffer: Buffer,
```

Comment:
we should not pass around the `startBuffer`, but create an initial one in `createAggregationBuffer`